src/elexmodel/models/ConformalElectionModel.py

```diff
@@ -0,0 +1,194 @@
+        # want the expanded_fixed_effects from train_data to be used by conformalization_data and nonreporting_data
+        # in this function.
+        train_data_features = self.featurizer.featurize_fitting_data(train_data, add_intercept=self.add_intercept)
+        train_data_residuals = train_data[f"residuals_{estimand}"]
+        train_data_weights = train_data[f"last_election_results_{estimand}"]
+
+        # fit lower and upper model to training data. ECOS solver is better than SCS.
+        lower_qr = QuantileRegressionSolver(solver="ECOS")
+        self.fit_model(lower_qr, train_data_features, train_data_residuals, lower_bound, train_data_weights, True)
+
+        upper_qr = QuantileRegressionSolver(solver="ECOS")
+        self.fit_model(upper_qr, train_data_features, train_data_residuals, upper_bound, train_data_weights, True)
+
+        # apply to conformalization data. Conformalization bounds will later tell us how much to adjust lower/upper
+        # bounds for nonreporting data.
+        conformalization_data = reporting_units_shuffled[train_rows:].reset_index(drop=True)
+        # conformalization features will be the same as the features in train_data
+        conformalization_data_features = self.featurizer.featurize_heldout_data(conformalization_data)
+
+        # we are interested in f(X) - r
+        # since later conformity scores care about deviation of bounds from residuals
+        conformalization_lower_bounds = (
+            lower_qr.predict(conformalization_data_features) - conformalization_data[f"residuals_{estimand}"].values
+        )
+        conformalization_upper_bounds = conformalization_data[f"residuals_{estimand}"].values - upper_qr.predict(
+            conformalization_data_features
+        )
+
+        # save conformalization bounds for later
+        conformalization_data["upper_bounds"] = conformalization_upper_bounds
+        conformalization_data["lower_bounds"] = conformalization_lower_bounds
+
+        # apply lower/upper models to nonreporting data
+        # now the features of the nonreporting_units will be the same as the train_data features
+        # they might differ slightly from the features used when fitting the median prediction
+        nonreporting_units_features = self.featurizer.featurize_heldout_data(nonreporting_units)
+        nonreporting_lower_bounds = lower_qr.predict(nonreporting_units_features)
+        nonreporting_upper_bounds = upper_qr.predict(nonreporting_units_features)
+
+        return PredictionIntervals(nonreporting_lower_bounds, nonreporting_upper_bounds, conformalization_data)
+
+    def get_all_conformalization_data_unit(self):
+        """
+        Returns conformalization data at the unit level
+        Conformalization data is adjustment from estimated % change from baseline
+        """
+        pass
+
+    def get_all_conformalization_data_agg(self):
+        """
+        Returns conformalization data at the aggregate level
+        """
+        pass
```

Comment:
Why are these two methods just pass-throughs? 🤔

Reply:
same as above. will change to NotImplemented